src/ZuluSCSI_tape.cpp

```diff
@@ -26,15 +29,624 @@
+    tap_record_t record;
+    uint8_t dummy_buffer[1];  // Small buffer for record parsing
+
+    for (uint32_t i = 0; i < count; i++) {
+        tap_result_t result = tapReadRecordForward(img, record, dummy_buffer, sizeof(dummy_buffer));
+
+        if (result == TAP_END_OF_TAPE) {
+            return TAP_END_OF_TAPE;
+        }
+
+        if (result == TAP_ERROR) {
+            return TAP_ERROR;
+        }
+
+        if (filemarks) {
+            // Spacing filemarks - stop when we hit one
+            if (result == TAP_FILEMARK) {
+                break;
+            }
+        } else {
+            // Spacing records - filemarks are also records for spacing purposes
+            // Continue until count reached or end of tape
+        }
+    }
+
+    return TAP_OK;
+}
+
+// Space backward by records or filemarks
+tap_result_t tapSpaceBackward(image_config_t &img, uint32_t count, bool filemarks) {
+    tap_record_t record;
+    uint8_t dummy_buffer[1];
+
+    for (uint32_t i = 0; i < count; i++) {
+        tap_result_t result = tapReadRecordBackward(img, record, dummy_buffer, sizeof(dummy_buffer));
+
+        if (result == TAP_BEGINNING_OF_TAPE) {
+            return TAP_BEGINNING_OF_TAPE;
+        }
+
+        if (result == TAP_ERROR) {
+            return TAP_ERROR;
+        }
+
+        if (filemarks) {
+            // Spacing filemarks - stop when we hit one
+            if (result == TAP_FILEMARK) {
+                break;
+            }
+        }
+    }
+
+    return TAP_OK;
+}
+
+// Rewind to beginning of tape
+tap_result_t tapRewind(image_config_t &img) {
+    img.tape_pos = 0;
+    return TAP_OK;
+}
+
+// State tracking for .TAP format writes
+struct tap_transfer_t {
+    uint32_t record_length;     // Total length of record being written
+    uint32_t bytes_written;     // Bytes written so far
+    bool is_fixed_mode;         // Fixed vs variable block mode
+    bool header_written;        // Has 4-byte header been written?
+    uint32_t file_pos_start;    // File position at start of record
+
+    // Fields for data transfer management (similar to g_disk_transfer)
+    uint8_t* buffer;            // Buffer for data transfer
+    uint32_t bytes_scsi;        // Total SCSI buffer size
+    uint32_t bytes_sd;          // Bytes processed so far
+    uint32_t bytes_scsi_started; // Bytes started in SCSI transfer
+    uint32_t sd_transfer_start; // Start position for SD transfer
+    int parityError;            // Parity error flag
+};
+
+static tap_transfer_t g_tap_transfer;
+
+#ifdef PREFETCH_BUFFER_SIZE
+static struct {
+    uint8_t buffer[PREFETCH_BUFFER_SIZE];
+    uint32_t position;          // Tape position for prefetch data
+    uint32_t bytes;             // Bytes available in prefetch buffer
+    uint8_t scsiId;             // SCSI ID for this prefetch
+} g_tape_prefetch;
+#endif
+
+// Check if the tape file is in .TAP format
+bool tapIsTapFormat(image_config_t &img) {
```

Comment:
Because `tapIsTapFormat()` is called for every transfer, I think it would be cleaner to check a predefined `bool` instead of parsing the filename every time. This can be added to `image_config_t`, for example `bool tape_is_tap_format;`.

When the image is first opened, I think it would be good to run a `tapReadRecordForward()` to verify the file format, and print an informative message for the user. Most ZuluSCSI image types do not require special formatting, so it can cause confusion if image loading is ok but transfers immediately fail.

Reply:
File name checking is probably fine, format errors should be returned to host and will appear as if uninitialized tape.